File: src/api/cost.ts
import { invokeWithError } from './invoke'

export interface RequestLog {
  id: string
  provider_id: string | null
  model: string | null
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  cache_creation_tokens: number
  total_cost_cny: number
  latency_ms: number | null
  first_token_ms: number | null
  status_code: number | null
  error_message: string | null
  session_id: string | null
  is_streaming: boolean
  created_at: number
}

export interface DailyUsage {
  date: string
  provider_id: string | null
  model: string | null
  request_count: number
  input_tokens: number
  output_tokens: number
  total_cost_cny: number
}

export interface CostSummary {
  total_requests: number
  total_input_tokens: number
  total_output_tokens: number
  total_cost_cny: number
  by_provider: { provider_id: string; provider_name: string; request_count: number; total_cost_cny: number }[]
  by_model: { model: string; request_count: number; input_tokens: number; output_tokens: number; total_cost_cny: number }[]
}

export interface DateRange {
  start: string
  end: string
}

export interface LogFilter {
  limit?: number
  offset?: number
}

/** 任务级指标（task_runs 聚合：成功率 / 耗时分布 / 成本 / 错误分类） */
export interface TaskStats {
  total_tasks: number
  success_count: number
  error_count: number
  cancelled_count: number
  /** 成功率 0~1（成功 / 非取消任务） */
  success_rate: number
  p50_ms: number | null
  p95_ms: number | null
  avg_duration_ms: number | null
  total_cost_cny: number
  total_input_tokens: number
  total_output_tokens: number
  /** 错误分类分布（kind -> 次数，按次数倒序） */
  by_error_kind: { kind: string; count: number }[]
}

/** 单次任务运行记录（trace 明细） */
export interface TaskRun {
  id: string
  conversation_id: string
  project_id: string
  provider_id: string | null
  model: string | null
  /** success | incomplete | error | cancelled */
  status: string
  error_kind: string | null
  error_message: string | null
  tool_rounds: number
  retry_count: number
  input_tokens: number
  output_tokens: number
  cost_cny: number
  duration_ms: number
  started_at: number
  finished_at: number
}

export const getCostSummary = (range: DateRange) => invokeWithError<CostSummary>('get_cost_summary', { range })
export const getRequestLogs = (filter: LogFilter) => invokeWithError<RequestLog[]>('get_request_logs', { filter })
export const getDailyUsage = (range: DateRange) => invokeWithError<DailyUsage[]>('get_daily_usage', { range })
/** 任务级指标聚合；project_id 为空 = 全局，days 缺省 30 */
export const getTaskStats = (projectId?: string, days = 30) =>
  invokeWithError<TaskStats>('get_task_stats', { projectId: projectId ?? '', days })

/** 最近任务列表；project_id 为空 = 全局；status 可选过滤（success/incomplete/error/cancelled） */
export const getTaskRuns = (projectId?: string, status?: string, limit = 20) =>
  invokeWithError<TaskRun[]>('get_task_runs', { projectId: projectId ?? '', status: status ?? '', limit })

/** 单 Provider 预算状态：当日/当月已用 + 日/月上限 */
export interface BudgetStatus {
  provider_id: string | null
  used_today_cny: number
  used_month_cny: number
  daily_limit_cny: number | null
  monthly_limit_cny: number | null
}

/** 全部 Provider 预算概览：单 Provider 列表 + 跨 Provider 汇总 */
export interface AllBudgetStatus {
  providers: BudgetStatus[]
  used_today_cny: number
  used_month_cny: number
  daily_limit_cny: number | null
  monthly_limit_cny: number | null
}

/** 查询某 Provider 预算状态；provider_id 缺省 = 跨 Provider 汇总 */
export const getBudgetStatus = (providerId?: string) =>
  invokeWithError<BudgetStatus>('get_budget_status', { providerId: providerId ?? null })

/** 全部 Provider 预算概览（成本页顶部预算总览卡片用） */
export const getAllBudgetStatus = () => invokeWithError<AllBudgetStatus>('get_all_budget_status')

File: src-tauri/src/commands/cost.rs
use tauri::State;
use crate::db::{
    models::{CostSummary, DailyUsage, RequestLog, TaskStats},
    queries, DbState,
};
use crate::services::budget::{self, BudgetStatus};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct DateRange {
    pub start: String,
    pub end: String,
}

#[derive(Debug, Deserialize)]
pub struct LogFilter {
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

#[tauri::command]
pub fn get_cost_summary(db: State<DbState>, range: DateRange) -> Result<CostSummary, String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;

    let daily = queries::get_daily_usage(&conn, &range.start, &range.end)
        .map_err(|e| e.to_string())?;

    let mut total_requests: i64 = 0;
    let mut total_input: i64 = 0;
    let mut total_output: i64 = 0;
    let mut total_cost: f64 = 0.0;

    for d in &daily {
        total_requests += d.request_count;
        total_input += d.input_tokens;
        total_output += d.output_tokens;
        total_cost += d.total_cost_cny;
    }

    // 按模型分组统计（请求日志明细聚合）
    let by_model = queries::get_cost_by_model(
        &conn,
        date_start_ts(&range.start),
        date_end_ts(&range.end),
    )
    .map_err(|e| e.to_string())?;

    Ok(CostSummary {
        total_requests,
        total_input_tokens: total_input,
        total_output_tokens: total_output,
        total_cost_cny: total_cost,
        by_provider: vec![],
        by_model,
    })
}

/// 日期字符串（YYYY-MM-DD）→ 当天 00:00 的秒级时间戳（UTC，与日志 created_at 对齐）
fn date_start_ts(date: &str) -> i64 {
    chrono::NaiveDate::parse_from_str(date, "%Y-%m-%d")
        .ok()
        .and_then(|d| d.and_hms_opt(0, 0, 0))
        .map(|dt| dt.and_utc().timestamp())
        .unwrap_or(0)
}

/// 日期字符串（YYYY-MM-DD）→ 当天 23:59:59 的秒级时间戳
fn date_end_ts(date: &str) -> i64 {
    chrono::NaiveDate::parse_from_str(date, "%Y-%m-%d")
        .ok()
        .and_then(|d| d.and_hms_opt(23, 59, 59))
        .map(|dt| dt.and_utc().timestamp())
        .unwrap_or(i64::MAX)
}

#[tauri::command]
pub fn get_request_logs(db: State<DbState>, filter: LogFilter) -> Result<Vec<RequestLog>, String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    let limit = filter.limit.unwrap_or(50);
    let offset = filter.offset.unwrap_or(0);
    queries::get_request_logs(&conn, limit, offset).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_daily_usage(db: State<DbState>, range: DateRange) -> Result<Vec<DailyUsage>, String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    queries::get_daily_usage(&conn, &range.start, &range.end).map_err(|e| e.to_string())
}

/// 任务级指标聚合（成功率 / P50 / P95 / 成本 / 错误分布）；project_id 为空 = 全局
#[tauri::command]
pub fn get_task_stats(db: State<DbState>, project_id: Option<String>, days: Option<i64>) -> Result<TaskStats, String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    let project_id = project_id.unwrap_or_default();
    let days = days.unwrap_or(30).clamp(1, 365);
    queries::get_task_stats(&conn, &project_id, days).map_err(|e| e.to_string())
}

/// 最近任务列表（trace 明细）：project_id 为空 = 全局；status 可选过滤；limit 缺省 20
#[tauri::command]
pub fn get_task_runs(
    db: State<DbState>,
    project_id: Option<String>,
    status: Option<String>,
    limit: Option<i64>,
) -> Result<Vec<crate::db::models::TaskRun>, String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    let limit = limit.unwrap_or(20).clamp(1, 200);
    queries::list_task_runs(&conn, &project_id.unwrap_or_default(), status.as_deref(), limit)
        .map_err(|e| e.to_string())
}

/// 预算状态查询（前端成本页展示）：provider_id 为 None 时跨 Provider 汇总
#[tauri::command]
pub fn get_budget_status(db: State<DbState>, provider_id: Option<String>) -> Result<BudgetStatus, String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    Ok(budget::budget_status(&conn, provider_id.as_deref()))
}

/// 全部 Provider 的预算状态（成本页"全部"概览用）
#[derive(Debug, serde::Serialize)]
pub struct AllBudgetStatus {
    pub providers: Vec<BudgetStatus>,
    /// 跨 Provider 汇总：日已用、月已用、合计日上限、合计月上限
    pub used_today_cny: f64,
    pub used_month_cny: f64,
    pub daily_limit_cny: Option<f64>,
    pub monthly_limit_cny: Option<f64>,
}

#[tauri::command]
pub fn get_all_budget_status(db: State<DbState>) -> Result<AllBudgetStatus, String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    // 拉所有 provider 的预算使用情况
    let mut stmt = conn
        .prepare("SELECT id, limit_daily_cny, limit_monthly_cny FROM providers WHERE enabled = 1")
        .map_err(|e| e.to_string())?;
    let rows: Vec<(String, Option<f64>, Option<f64>)> = stmt
        .query_map([], |r| Ok((r.get(0)?, r.get(1)?, r.get(2)?)))
        .map_err(|e| e.to_string())?
        .filter_map(|r| r.ok())
        .collect();
    drop(stmt);

    let mut providers = Vec::new();
    let mut sum_today = 0.0;
    let mut sum_month = 0.0;
    let mut sum_daily: Option<f64> = Some(0.0);
    let mut sum_monthly: Option<f64> = Some(0.0);
    for (pid, dl, ml) in &rows {
        let s = budget::budget_status(&conn, Some(pid));
        providers.push(s);
        sum_today += providers.last().unwrap().used_today_cny;
        sum_month += providers.last().unwrap().used_month_cny;
        if let Some(d) = dl {
            sum_daily = sum_daily.map(|x| x + d);
        } else {
            sum_daily = None;
        }
        if let Some(m) = ml {
            sum_monthly = sum_monthly.map(|x| x + m);
        } else {
            sum_monthly = None;
        }
    }
    // 修正：只要有一个 provider 没设上限，合计上限就为 None
    if rows.iter().any(|(_, d, _)| d.is_none()) {
        sum_daily = None;
    }
    if rows.iter().any(|(_, _, m)| m.is_none()) {
        sum_monthly = None;
    }
    Ok(AllBudgetStatus {
        providers,
        used_today_cny: sum_today,
        used_month_cny: sum_month,
        daily_limit_cny: sum_daily,
        monthly_limit_cny: sum_monthly,
    })
}
